### PR TITLE
Removal of _value parameter in ensure_account_can_withdraw

### PR DESCRIPTION
**Pull request by thiolliere**
related to https://github.com/paritytech/substrate/issues/1962

as pointed by @xlc [here](https://github.com/paritytech/substrate/pull/1921/files#r264077649) we don't check if account can withdraw the amount for the TransactionPayment reason.

so it fix it, improve ensure_account_can_withdraw syntax and remove _amount unused parameter from it.

I am not very sure of this, I may be wrong if TransactionPayment reason is considered a subset of Transfer reason.

### CODE REVIEW

File: srml/balances/src/lib.rs [hunk 1]
```diff
@@ -370,26 +374,26 @@ impl<T: Trait> Module<T> {
...
 	/// Returns `Ok` iff the account is able to make a withdrawal of the given amount
 	/// for the given reason.
-	/// 
+	///
 	/// `Err(...)` with the reason why not otherwise.
 	pub fn ensure_account_can_withdraw(
 		who: &T::AccountId,
-		_amount: T::Balance,
-		reason: WithdrawReason,
+		reasons: impl Into<WithdrawReasons>,
```
Comment (xlc) [inline srml/balances/src/lib.rs]: I never encounter this syntax before. Can I know where I can read more about this syntax?

Comment (thiolliere) [reply srml/balances/src/lib.rs]: in parameter as far as I know it is same as if I have a generic in function `T: Into<WithdrawReasons>` and then `reasons: T` I don't have that much material maybe try to google `impl trait`

File: srml/balances/src/lib.rs [hunk 2]
```diff
@@ -302,7 +302,11 @@ impl<T: Trait> Module<T> {
 		if would_create && value < Self::existential_deposit() {
 			return Err("value too low to create account");
 		}
-		Self::ensure_account_can_withdraw(transactor, value, WithdrawReason::Transfer, new_from_balance)?;
+		Self::ensure_account_can_withdraw(
+			transactor,
+			WithdrawReason::Transfer | WithdrawReason::TransactionPayment,
```
Comment (gavofyork) [inline srml/balances/src/lib.rs]: This is not a transaction payment. Transaction payments happen specifically for inclusion of a transaction in the chain.

File: srml/balances/src/lib.rs (hunk 1 above)
Comment (gavofyork) [reply srml/balances/src/lib.rs]: Incorrect - there should be just one `WithdrawReason` for any single withdraw,

**Comment by gavofyork** [comment]
Removal of `_value` parameter is reasonable. However the changes are in error.

**Comment by xlc** [comment]
Seems like me and @thiolliere have some misunderstanding of the meaning of transaction payment. @gavofyork can you provide some definitions of it?

To my understanding, transaction payment, is the transaction fee user have to pay to have the transaction included in the chain.
This can be made up with two parts:
1. fee of inclusion of the the transaction, which is calculated with a base fee + transaction size * byte fee
2. fee of additional storage / computation usage, in balance transfer case is account creation fee (addition storage usage) and transfer fee (storage read and write). In smart contract case, it is gas fee.

If my understanding is wrong, that only (1) is counted as transaction payment, then what is the (2) part? It is a fee charged to user. It is not value transferred to another account. Should there be another `WithdrawReason`?

File: srml/balances/src/lib.rs (hunk 1 above)
Comment (xlc) [reply srml/balances/src/lib.rs]: Let's say I want to create a runtime module that allow user to reserve and transfer in a single runtime call. How can do it? Call `ensure_account_can_withdraw` twice with different reasons before all the withdraws will not give the correct result.

Comment (thiolliere) [reply srml/balances/src/lib.rs]: actually calling 
```rust
ensure_account_can_withdraw(who, Reserve | Transfer, new_free_balance)
```
is equivalent to
```rust
ensure_account_can_withdraw(who, Transfer, new_free_balance)
ensure_account_can_withdraw(who, Reserve, new_free_balance)
```

**Comment by thiolliere** [comment]
Yes that was thought indeed, then maybe (2) is actually just a Transfer reason

**Comment by gavofyork** [comment]
So `WithdrawReason::Transfer` is to be read "withdrawing from an account in the duty of executing the `transfer` function".

It doesn't relate to the balance of the total withdrawal that lands in the transfer destination's account.

**Pull request closed without merging**